Complete rewrite, dynamic resources, <require from> fixes

niieani opened the pull request:
Fixes: 
- https://github.com/aurelia/webpack-plugin/issues/8
- https://github.com/aurelia/webpack-plugin/issues/11

martijnboland commented:
Wow, impressive!

atsu85 commented on src/resolve-template.js:
```diff
@@ -0,0 +1,90 @@
+
+/**
+ * Generates key-value pairs or <require from="paths"> and their corresponding files.
+ */
```
nice, code from this method 
_(iterating over all html files and checking that `from` attributes of `require` tags reference files, that exist in the project)_
could be taken as a base for implementing [Build-time validation (linter) for Aurelia application](https://github.com/aurelia/tools/issues/31), namely for
> typos with incorrect `<require from="./not/found">` element path

niieani replied:
I'm currently working on even more extensive scraping of requires, such as `view-model="file"` and `view="file.html"`. @EisenbergEffect Are there any more ways of importing files/modules dynamically from the view?

EisenbergEffect replied:
Those are the ways that are defined by the core. It's possible for developers to invent new ones ;)

Once this is built, it might also be a good idea to extract the core logic so its independent from webpack, as @atsu85 says, there are a lot of other things we could do with this probably. We don't need to worry about that just yet, but lets keep it in mind for the future.

atsu85 replied:
I guess You mean `compose` element with those attributes? That is a good idea, but I'm not sure how often those attributes contain plain strings in real life as opposed to expressions that You probably can't detect at build-time.

niieani replied:
@EisenbergEffect @atsu85 Indeed it is possible to invent new ones, plus it is possible to generate paths dynamically. Something like this cannot be supported by statically analyzing files. For that purpose I'm implementing an additional way to declare external dependencies and for Aurelia plugins to declare internal dependencies by listing those additional resources in the `package.json` file. You'd add something like:

```
{
   ...
   "aurelia": {
      "resources": [ 'dist/some-resource.js', 'dist/another.html', 'dist/another-without-extension' ]
    }
}
```
And the Webpack plugin will read that and know to include those files in the bundle.

niieani commented:
I've updated the pull request with some more features:
Since there are cases that cannot be supported by statically analyzing files (e.g. it is possible to generate require paths dynamically), I've implemented an additional way to declare external dependencies and for Aurelia plugins to declare internal dependencies by listing those additional resources in the `package.json` file. You'd add something like:

```js
{
   ...
   "aurelia": {
      "resources": [
        "./src/some-resource.js",
        "./src/another.html",
        "./src/another-without-extension",
        { "path": "external-module/file.html", async: true, bundle: 'some-bundle' }
      ]
   }
}
```

And the Webpack plugin will read that and know to include those files in the bundle.
It'll be intelligent enough to resolve any further dependencies that those files then require statically (regardless if they are JavaScript files or additional HTML resources). 
As you can see it is also possible to specify that certain files should be lazy-loaded or bundled separately (if the bundle is specified in the configuration file). To achieve the same directly from templates, add the parameters `async` or `bundle="bundle-name"` to either the `<require>` tag or to `<compose>` tag, when composing with static names. E.g.:

```html
<require from="./some-file.html" async bundle="other-bundle">
```

EisenbergEffect commented:
@ahmedshuhel Please review this. Some of what @niieani is doing for Webpack is similar to what I've talked to you about for supporting plugin bundling with system-builder. If we could standardize on a format for the package json then we could make webpack, system-builder and require.js all work with this automatically, which would be a great experience for our community. Please review this and let's discuss here.

ahmedshuhel commented:
Reviewing the PR. I will post my findings.

niieani commented:
Small update. I've deprecated `includeSubModules` and cleaned up the code @ `index` cause it was really hard to read. At the moment it is necessary to include this in `package.json` for an Aurelia project:

```json
  "aurelia": {
    "resources": [
      "aurelia-templating-resources/compose",
      "aurelia-templating-resources/if",
      "aurelia-templating-resources/with",
      "aurelia-templating-resources/repeat",
      "aurelia-templating-resources/show",
      "aurelia-templating-resources/hide",
      "aurelia-templating-resources/replaceable",
      "aurelia-templating-resources/sanitize-html",
      "aurelia-templating-resources/focus",
      "aurelia-templating-resources/compile-spy",
      "aurelia-templating-resources/view-spy",
      "aurelia-templating-resources/binding-mode-behaviors",
      "aurelia-templating-resources/throttle-binding-behavior",
      "aurelia-templating-resources/debounce-binding-behavior",
      "aurelia-templating-resources/signal-binding-behavior",
      "aurelia-templating-resources/update-trigger-binding-behavior",
      "aurelia-templating-router/router-view",
      "aurelia-templating-router/route-href"
    ],
    "moduleRootOverride": {
      "aurelia-templating-resources": "dist/es2015",
      "aurelia-templating-router": "dist/es2015"
    }
  },
```

Perhaps once we have agreed on the format and this is merged we can declare those resources in the package.json of those respectful packages.

cehoffman commented:
Hey @niieani, much thanks for trying to bring the webpack loader to a better state with support for 2.0. I've tried  your fork of aurelia-webpack-plugin with the above inclusions to my `package.json` and the webpack configuration files form aurelia/skeleton-navigation#440. Obviously it looks like the skeleton navigation PR has become a little out of sync with the changes on this PR.

It looks like there is still an issue with webpack using the `commonjs` files for the aurelia packages despite the aliases form the navigation PR, `moduleType` argument to `AureliaWebpackPlugin`, and the `aurelia` property in `package.json`. I know the plugin is trying to use the `es2015` sources because a build results in many lines like this which I traced to the plugin.

~~~
[package.json] required "./aurelia-templating-resources/compose.js" from "node_modules/aurelia-templating-resources/dist/es2015/compose.js".
[package.json] required "./aurelia-templating-resources/compose" from "node_modules/aurelia-templating-resources/dist/es2015/compose.js".
[package.json] required "./aurelia-templating-resources/dist/es2015/if" from "node_modules/aurelia-templating-resources/dist/es2015/if.js".
[package.json] required "./aurelia-templating-resources/if.js" from "node_modules/aurelia-templating-resources/dist/es2015/if.js".
[package.json] required "./aurelia-templating-resources/if" from "node_modules/aurelia-templating-resources/dist/es2015/if.js".
[package.json] required "./aurelia-templating-resources/dist/es2015/with" from "node_modules/aurelia-templating-resources/dist/es2015/with.js".
[package.json] required "./aurelia-templating-resources/with.js" from "node_modules/aurelia-templating-resources/dist/es2015/with.js".
~~~

Yet when I look at the built `aurelia.bundle.js` the sourceMap comments all reference the commonjs path. What sparked me to notice this was the output running the webpack-dev-server. It has several lines like this at the end of the build.

~~~
  [107] ./~/aurelia-templating-resources/dist/commonjs/with.js 1.73 kB {2} [built]
       [158] 0ms -> [55] 1001ms -> factory:176ms building:112ms dependencies:1ms = 1290ms
  [108] ./~/aurelia-templating-router/dist/commonjs/route-href.js 2.53 kB {2} [built]
       [158] 0ms -> [63] 1017ms -> factory:167ms building:181ms dependencies:2ms = 1367ms
  [109] ./~/aurelia-templating-router/dist/commonjs/route-loader.js 2.69 kB {2} [built]
       [158] 0ms -> [63] 1017ms -> factory:167ms building:162ms dependencies:13ms = 1359ms
  [110] ./~/aurelia-templating-router/dist/commonjs/router-view.js 6.76 kB {2} [built]
       [158] 0ms -> [63] 1017ms -> factory:167ms building:178ms dependencies:2ms = 1364ms
~~~

At the start of the build though it looks like it analyzes the es2015 sources

~~~
  [177] ./~/aurelia-templating-resources/dist/es2015/with.js 1.51 kB {0} [optional] [built]
       [158] 0ms -> [41] 819ms -> [160] 1129ms -> factory:128ms building:446ms dependencies:0ms = 2522ms
  [178] ./~/aurelia-templating-router/dist/es2015/route-href.js 1.88 kB {0} [optional] [built]
       [158] 0ms -> [41] 819ms -> [160] 1129ms -> factory:128ms building:448ms dependencies:4ms = 2528ms
  [179] ./~/aurelia-templating-router/dist/es2015/router-view.js 6.43 kB {0} [optional] [built]
       [158] 0ms -> [41] 819ms -> [160] 1129ms -> factory:129ms building:454ms dependencies:1ms = 2532ms
~~~

My guess is this might be part of that router-view problem you were seeing. One piece of code is using the ES2015 emitted code and the other the commonjs emitted code.

Anyway, I'm posting this here since I just spent the last 4-5 hours starting a new Aurelia project and dove headfirst into a brick wall trying to use Webpack instead of JSPM. I nearly abandoned it until I found your PRs. To be fair to Webpack, the problems are the same ones I had when trying to use JSPM for the first time. I just know how to solve them in JSPM now after equal amounts of head banging when JSPM support was just as fleshed out as Webpack is now in Aurelia.

niieani commented:
Hi @cehoffman ! In my testing yesterday, I have discovered more bugs related to es2015 imports usage in Webpack, which seem to do with how ES7 decorators are transpiled in the es2015 builds. (DOM from aurelia-pal doesn't seem to be dynamically imported via `@inject`, and there were problems with `@autoinject`). I think the best solution is to simply use `commonjs` builds of Aurelia for now, especially since tree-shaking for Aurelia packages doesn't give too much of a reduction in file-size. Tree shaking will still work for external packages and internal code.
I'll try to post an updated webpack skeleton today.

niieani commented:
@EisenbergEffect I've pushed my fixes to both the **webpack skeleton** and this plugin. This is tested really thoroughly now and a lot of use cases are covered (including dynamic imports and bundling). I've removed the prepare release commits here as well. There was no build system on this package previously (was pure ES5), but my changes have introduced one (it is based on other aurelia plugins though), so before you publish, you'll need to build. In order to show users how to add modules that are to be lazy-loaded by Aurelia, I've added a feature to the skeleton so that `users` module and its dependencies are bundled to an external file.

btw. I was able to fix es2015 imports too, I've made a small test to see what is the real benefit in tree-shaking aurelia's packages. The gain is about 35KB. 
Base Aurelia packages used by the skeleton bundled for production come to about 350KB. After minification and gzipping that's a bit less than 80KB (Aurelia only). 
Bundled with Bluebird, jQuery, Bootstrap and Skeleton Navigation - all minified and gzipped the website is only 153 KB (fonts not included). Webpack makes it possible to do conditional module imports, which is useful for polyfilling only if necessary. That could mean savings of 50 KB in case of modern browsers. I've added an example to the skeleton of how to polyfill fetch-client dynamically.

@cehoffman You should be able to try the new skeleton once all the changes are merged and published. I believe I've got all the major bugs eliminated, I feel that at this this is much more solid than JSPM.

EisenbergEffect commented:
@niieani Nice work! I'm having @ahmedshuhel also do a review before we merge. But, seems pretty solid to me.

niieani commented:
@ahmedshuhel any progress? ⭐

EisenbergEffect commented:
@niieani I'm wondering if we could put "resources" inside of a "build" section of the "aurelia" section. I'm thinking that we might use that section for a few other things to support other tools in the future. Does that sound reasonable? Also, can you explain the purpose of "moduleRootOverride"? Thanks!

niieani commented:
@EisenbergEffect final design of package.json is up to you. The structure I put forth here ( `package.json -> aurelia -> resouces[] -> resource {string or object}` ) is just a proposition. We could even create something like an `aurelia.json` configuration file. It's all possible. I just don't like having a million files in the root dir laying around, so I thought package.json might be a good place to integrate this, since Aurelia's packages already have an "aurelia" object defined there.

I should point out that I've made it possible to declare resources in such a way that Webpack will separate them out into separate, lazy-loaded bundles, which is demonstrated in the new webpack-skeleton [here](https://github.com/niieani/skeleton-navigation-typescript/blob/new-webpack/skeleton-typescript-webpack/package.json#L76-L81). In the example case, `/src/users.ts`, `/src/users.html` and all of their dependencies -- in this case "/src/blur-image.ts" and the "aurelia-fetch-client" module - will be automatically separated out into its own file and loaded on demand, once the user loads the /users route.

Also, resources which are declared inside of .html Views do not need to be redeclared in package.json.

As for `moduleRootOverride`, it is used to signify that any import from a given package should have its `root` amended to a given path. In the example, we simply ensure that both templating-resources and templating-router use the es2015 builds instead of the commonjs ones that would be inferred by "adjacent to main" logic (step nr 5 below). 

By default the logic of importing a non-local path is as follows (if the file doesn't exist, we continue):

Example: `<require from="bootstrap/bootstrap.css">`

1. if the require is inside the project: try from local src root (defined by user in the Webpack config file): 
##### resolves to
   ```
   __project_folder__/src/bootstrap/bootstrap.css
   ```

2. if the `root` property is set for a given import in (first project's package.json, then bootstrap's package.json), try to resolve that location:
   ```js
     "aurelia": {
       "resources": [
         { path: "bootstrap/bootstrap.css", root: "custom/path/here" }
       ]
     },
   ```
##### resolves to:
   ```
   node_modules/bootstrap/custom/path/here/bootstrap.css
   ```

3. if the `moduleRootOverride` is set in the module (project's package.json, then bootstrap's package.json), try that:
   ```
       "moduleRootOverride": {
         "bootstrap": "some/other/path"
       }
   ```
##### resolves to:
   ```
   node_modules/bootstrap/some/other/path/bootstrap.css
   ```

4. try real requested path:
   ```
   node_modules/bootstrap/bootstrap.css
   ```

5. try subdirectory adjacent to package's `main`:
   e.g. given bootstrap defines `main` as "dist/bootstrap.js":
   ##### resolves to:
   ```
   node_modules/bootstrap/dist/bootstrap.css
   ```

In reality there are even more steps if we requested the file without an extension, because for every step we try to find a file with `.js`, then `.js` and possibly a matching `.html`.

EisenbergEffect commented:
I'd like to keep things in the package.json file, at least for plugins. That seems reasonable. I'd also like ot move the resources under a `build` key. We might add more things there in the future and I'd like to keep that grouped together. I'm thinking of some other settings that might go there to enable other build systems.

It seems you are tackling two different but related things with resources though. One is the ability to properly bundle plugins and the other relates to application bundling. I'm not 100% confident that the second scenario belongs under a resources key. It seems a bit out of place.

I wouldn't be so picky about this, but I'm trying to see if we can use this information for another bundling solution as well, so I just want to make sure its clean and potentially webpack agnostic if possible, or at least organize it so things are separated along those lines.

Thoughts?

niieani commented:
All of the properties I've specified could theoretically be consumed by other bundlers, none of this is Webpack-specific.

As for your comment on how this is used to bundle plugins and application bundling -- I've thought about it before and I've reached the conclusion that there's no way to somehow separate these things out. It would be an artificial separation.

Every module is scoped, and thus cannot influence the way the project is bundled outside of its own dependencies. My reasoning is:

A locally defined ValueConverter (in a separate file) should be definable in the same way as a plugin's ValueConverter. After all, you might want to separate it out, or lazy-load with other resources. 

Plus, in most cases plugins can define their own resources, which means all the user would have to do is include that module as a resource and my dependency-resolution logic will include all of its declared resources (and recursively, all of their required resources). This means we won't need [all this](https://github.com/niieani/skeleton-navigation-typescript/blob/new-webpack/skeleton-typescript-webpack/package.json#L82-L100) in the skeleton - if we declared those resources in their respectable repos. Finally [references to other modules](https://github.com/niieani/skeleton-navigation-typescript/blob/new-webpack/skeleton-typescript-webpack/package.json#L101-L107) could be declared as resources of `aurelia-framework` itself, and then that would be the only resource that the skeleton would need to explicitly require (though we could even bake one that into the plugin).

EisenbergEffect commented:
I think the thing that bothers me is this the lazy bundle config details. That would only be used in an app's package.json right? not in a plugin? That's the part that doesn't seem to fit with other bundling technologies. In other words, I want to make sure that what plugins put in their package.json files is compatible with other tech. If we want to support additional webpack specific build instructions in package.json, I think that should be restricted to an app's package file. Does that make sense?

niieani commented:
@EisenbergEffect If I understand you correctly, you'd like to add logic that prohibits modules from setting their requires as lazy or separated out to an external bundle, right? A plugin could arguably utilize bundle definitions or lazy loading to separate out parts (e.g. a UI plugin could separate out "button" and "table") - but in all fairness, it makes more sense to me for the user to do that explicitly, if he wanted to. 

I'm not familiar with how JSPM does bundling, but I'd imagine it supports separating files/modules into their own bundles and lazy-loading on need-to-have basis? In the JSPM skeleton I see the bundles are defined manually in `build/bundles.js`. I think these definitions could effectively replace that, making the bundling more seamless (and even switching between loaders/packers easier).

EisenbergEffect commented:
Yes, jspm can do that sort of thing. I'm thinking it probably just doesn't belong in a plugin, particularly since the plugin doesn't know what the bundles would be and it's usually the end developer that needs to decide how the app will be deployed. Mainly, I'd like to build some tooling on top of what you've got for require.js (and maybe system.js too...without jspm). I've already got part of it done. I think I can use most of what you've got too, perhaps even the same code. I just want to make sure I don't have to account for the bundle config inside of a plugin's package.json. That didn't make sense to me. I'm trust desiring to make sure I understand all your thinking here so that it can be accurately replicated in other places. If possible, I'd love to be able to use this metadata to support multiple options. That would be particularly important for plugin authors.

niieani commented:
@EisenbergEffect Then we're on the same page. I don't think plugins should define bundling metadata (laziness or bundle names) - that indeed doesn't make much sense - they should simply define their dependencies, should there be any - and the user can override this metadata with his own, should he want to (currently the top-most package always overrides the config - top-most being the users project).

EisenbergEffect commented:
Ok, excellent. We are in agreement. The final thing is the location of the data. I'd like it all under a sub key of the aurelia object in package.json. My thought was to put it under `build`. I'm open to other suggestions though. It could be `bundle` for example. Any thoughts on that?

niieani commented:
@EisenbergEffect I'm okay with either of those. I can do `build`. But tomorrow, it's going on 3 am here 😴 .

I'll add one more thing to what I wrote above: There might be times where the plugin doesn't want to define their dependencies in their package.json, but wants to leave that to the user. This is useful for componentized plugins, like UI plugins that can be used with only a subset of of their files (like - only the "button" component).

EisenbergEffect commented:
Sounds good.

niieani commented:
@EisenbergEffect Done.

EisenbergEffect commented:
Awesome work. The community is going to be very happy with this.
